sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataProducer.scala

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+
+import ai.rapids.cudf.Table
+
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * A GpuDataProducer produces data on the GPU. That data typically comes from other resources also
+ * held on the GPU that cannot be released until the iterator is closed and cannot be made
+ * spillable. This behaves like an Iterator but is not an Iterator because this cannot be used
+ * in place of an Iterator, especially in the context of an RDD where it would violate the
+ * semantics of the GpuSemaphore. Generally the lifetime of this should be entirely while the
+ * GpuSemaphore is held. It is "generally" because there are a few cases where for performance
+ * reasons if we know that no data is going to be produced, then we might not grab the semaphore
+ * at all.
+ *
+ * @tparam T what it is that we are wrapping
+ */
+abstract class GpuDataProducer[T] extends AutoCloseable {
+  /**
+   * Returns true if there is more data to be read or false if there is not.
+   */
+  def hasNext: Boolean
+
+  /**
+   * if hasNext returned true return the data. The reader is responsible for closing the
+   * returned value if it needs to be closed. If there is no more data to be read then
+   * an instance of NotSuchElementException should be thrown.
+   */
+  def next: T
+
+  /**
+   * Just like foreach on an Iterator
+   */
+  def foreach[U](func: T => U): Unit = {
+    while (hasNext) {
+      func(next)
+    }
```

Comment:
In cudf we had issue when not calling `next` before `hasNext` for the empty input file. In such cases, the valid output should be a table having 0 rows but with all (0 rows) columns in the file schema and we can only get such output table if we call `next` before `hasNext`:
```
do {
  func(next)
} while(hasNext)
```
I'm not sure if such output table is also desired in Spark.

Reply:
Wow that is totally not the order that I would expect the APIs to need to be called in. It is totally opposite of all java iterators, but OK. I will need to change it in a very different place than this because GpuDataProducer is abstract and says that it should operate like an iterator.

Reply:
Please make sure that the JNI API is clearly documented because like I said this totally is the opposite of java semantics.

Reply:
Is documenting this weird behavior the right answer?  This seems like behavior that should be fixed in cudf.  The entire point of `hasNext` is to be a predicate function to let the calling code know that it is valid to call `next`, as the method name implies.  Needing to special-case that for the first call is bizarre and error-prone, IMO.

Reply:
This is not just a specific issue with cudf, but a general logic issue. In the case of empty file, what should `hasNext` return? It should always return `false` as there is no data in the file (except metadata) to read. But what you expect is a table with empty columns, not a table without anything, right? So if you check `hasNext` and see a `false` then `next` will never be called and you never get a table with empty columns.

Reply:
Talked with Bobby and decided to workaround this in cudf, `hasNext` will return true exactly once.